File: packages/core/test/rivet/praxis/reviewer_independence.test.ts
import { describe, expect, test } from "bun:test"
import { BlindReviewerEngine, type ReviewPayload } from "../../../src/rivet/praxis/reviewer"
import {
  Revision,
  Scope,
  createObligationId,
  createEvidenceId,
} from "../../../src/rivet/types"

describe("Blind Reviewer Independence & Non-Inflation (I-08, I-14)", () => {
  test("Invariant I-08: Leaking implementer reasoning into review payload is rejected immediately", async () => {
    const engine = new BlindReviewerEngine()
    const payload: ReviewPayload = {
      goalDescription: "Fix security bug",
      obligationId: createObligationId(),
      scope: Scope.global("rivet", Revision.ZERO),
      diffContent: "--- a/lib.ts\n+++ b/lib.ts\n@@ -1 +1 @@\n-old\n+new",
      evidenceRefs: [createEvidenceId()],
      containsImplementerReasoning: true, // Leaked thought process
    }

    expect(engine.evaluateReview(payload)).rejects.toThrow(
      /Invariant I-08 violation/
    )
  })

  test("Invariant I-14: Duplicate reviews detect correlation and do not inflate confidence", async () => {
    const engine = new BlindReviewerEngine()
    const obligationId = createObligationId()
    const diff = "--- a/lib.ts\n+++ b/lib.ts\n@@ -1 +1 @@\n-old\n+new"

    const payload: ReviewPayload = {
      goalDescription: "Fix security bug",
      obligationId,
      scope: Scope.path("rivet", "lib.ts", Revision.ZERO),
      diffContent: diff,
      evidenceRefs: [createEvidenceId()],
      containsImplementerReasoning: false,
    }

    const firstVerdict = await engine.evaluateReview(payload)
    expect(firstVerdict.status).toBe("approved")
    if (firstVerdict.status === "approved") {
      expect(firstVerdict.confidence).toBe(0.95)
    }

    // Submit identical duplicate review
    const secondVerdict = await engine.evaluateReview(payload)
    expect(secondVerdict.status).toBe("approved")
    if (secondVerdict.status === "approved") {
      expect(secondVerdict.confidence).toBe(0.5) // Reduced confidence, no inflation
      expect(secondVerdict.rationale).toContain("duplicate review detected")
    }
  })

  test("Destructive commands in diff are rejected by blind reviewer", async () => {
    const engine = new BlindReviewerEngine()
    const payload: ReviewPayload = {
      goalDescription: "Fix bug",
      obligationId: createObligationId(),
      scope: Scope.global("rivet", Revision.ZERO),
      diffContent: "+++ b/run.sh\n+rm -rf /",
      evidenceRefs: [],
      containsImplementerReasoning: false,
    }

    const verdict = await engine.evaluateReview(payload)
    expect(verdict.status).toBe("rejected")
    if (verdict.status === "rejected") {
      expect(verdict.reasons[0]).toContain("Destructive operation")
    }
  })
})

File: legacy/rivet-rust/crates/praxis/tests/reviewer_independence_test.rs
use praxis::{BlindReviewerEngine, ReviewPayload, ReviewVerdict};
use rivet_types::*;

#[tokio::test]
async fn test_reviewer_rejects_implementer_reasoning_leakage_i08() {
    let engine = BlindReviewerEngine::new();

    let leaked_payload = ReviewPayload {
        goal_description: "Refactor auth middleware".into(),
        obligation_id: ObligationId::new(),
        scope: Scope::path("rivet", "src/auth/**", Revision::ZERO),
        diff_content: "+++ b/src/auth/mod.rs\n+pub fn check() {}".into(),
        evidence_refs: vec![],
        contains_implementer_reasoning: true, // Invariant I-08 breach
    };

    let result = engine.evaluate_review(&leaked_payload).await;
    assert!(result.is_err());
    let err_msg = result.unwrap_err().to_string();
    assert!(err_msg.contains("Invariant I-08"));
}

#[tokio::test]
async fn test_reviewer_detects_scope_breach_and_destructive_ops() {
    let engine = BlindReviewerEngine::new();

    let out_of_scope_payload = ReviewPayload {
        goal_description: "Refactor auth only".into(),
        obligation_id: ObligationId::new(),
        scope: Scope::path("rivet", "src/auth/**", Revision::ZERO),
        diff_content: "+++ b/src/billing/mod.rs\n+pub fn hijack() {}".into(),
        evidence_refs: vec![],
        contains_implementer_reasoning: false,
    };

    let verdict = engine.evaluate_review(&out_of_scope_payload).await.unwrap();
    assert!(matches!(verdict, ReviewVerdict::Rejected { .. }));

    if let ReviewVerdict::Rejected { reasons, .. } = verdict {
        assert!(
            reasons
                .iter()
                .any(|r| r.contains("outside declared write scope"))
        );
    }
}

#[tokio::test]
async fn test_reviewer_approves_valid_patch_and_handles_duplicate_i14() {
    let engine = BlindReviewerEngine::new();
    let obligation_id = ObligationId::new();

    let valid_payload = ReviewPayload {
        goal_description: "Refactor auth middleware".into(),
        obligation_id: obligation_id.clone(),
        scope: Scope::path("rivet", "src/auth/**", Revision::ZERO),
        diff_content: "+++ b/src/auth/mod.rs\n+pub fn check() -> bool { true }".into(),
        evidence_refs: vec![],
        contains_implementer_reasoning: false,
    };

    // First review: full confidence (0.95)
    let verdict1 = engine.evaluate_review(&valid_payload).await.unwrap();
    assert!(verdict1.is_approved());
    if let ReviewVerdict::Approved { confidence, .. } = verdict1 {
        assert!((confidence - 0.95).abs() < 0.01);
    }

    // Duplicate review: I-14 non-inflation (confidence 0.5)
    let verdict2 = engine.evaluate_review(&valid_payload).await.unwrap();
    assert!(verdict2.is_approved());
    if let ReviewVerdict::Approved { confidence, .. } = verdict2 {
        assert!((confidence - 0.5).abs() < 0.01);
    }
}
